app/controllers/api/v1/project_preferences_controller.rb

```
@@ -2,10 +2,34 @@ class Api::V1::ProjectPreferencesController < Api::ApiController
   include PreferencesController
 
   require_authentication :all, scopes: [:project]
+  extra_schema_actions :update_settings
   schema_type :json_schema
+  before_action :validate_records, only: [:update_settings]
+
+  def update_settings
+    upp = UserProjectPreference.find_or_initialize_by(project_id: params_for[:project_id],
+                                                      user_id: params_for[:user_id])
+    upp.settings.merge! params_for[:settings]
+    upp.save
```

This should be save! to raise and error on invalid record.